packages/react-combobox/src/utils/useOptionCollection.ts

```diff
@@ -1,91 +1,71 @@
-      } else {
-        return React.cloneElement(child);
-      }
-    }
-  });
-
-  return { keys, children: clonedChildren };
-}
+import type { OptionCollectionState, OptionValue } from './OptionCollection.types';
 
 /**
  * A hook for managing a collection of child Options
  */
-export const useOptionCollection = (children: React.ReactNode): OptionCollectionState => {
-  const optionData: React.MutableRefObject<OptionData> = React.useRef({});
-
-  const { collectionData, processedChildren } = React.useMemo(() => {
-    const { keys: optionKeys, children: clonedChildren } = getValidOptions(children);
+export const useOptionCollection = (): OptionCollectionState => {
+  const nodes = React.useRef<{ option: OptionValue; element: HTMLElement }[]>([]);
 
-    const getOptionAtIndex = (index: number) => {
-      const key = optionKeys[index];
-      return optionData.current[key];
-    };
-    const getIndexOfKey = (id: string) => optionKeys.indexOf(id);
-
-    const getOptionByKey = (key: string) => {
-      return optionData.current[key];
+  const collectionData = React.useMemo(() => {
+    const count = () => nodes.current.length;
+    const getOptionAtIndex = (index: number) => nodes.current[index]?.option;
+    const getIndexOfId = (id: string) => nodes.current.findIndex(node => node.option.id === id);
+    const getOptionById = (id: string) => {
+      const item = nodes.current.find(node => node.option.id === id);
+      return item?.option;
     };
 
     return {
-      collectionData: {
-        count: optionKeys.length,
-        getOptionAtIndex,
-        getIndexOfKey,
-        getOptionByKey,
-      },
-      processedChildren: clonedChildren,
+      count,
+      getOptionAtIndex,
+      getIndexOfId,
+      getOptionById,
     };
-  }, [children]);
+  }, []);
 
   const registerOption = React.useMemo(() => {
-    const register = (option: OptionValue) => {
-      if (option && option.key) {
-        optionData.current[option.key] = option;
+    const register = (option: OptionValue, element: HTMLElement) => {
+      const index = nodes.current.findIndex(node => {
+        if (!node.element || !element) {
+          return false;
+        }
+
+        if (node.option.id === option.id) {
+          return true;
+        }
+
+        // use the DOM method compareDocumentPosition to order the current node against registered nodes
+        return Boolean(node.element.compareDocumentPosition(element) === Node.DOCUMENT_POSITION_PRECEDING);
+      });
+
+      // do not register the option if it already exists
+      if (nodes.current[index]?.option.id !== option.id) {
+        const newItem = {
+          element,
+          option,
+        };
+
+        // If an index is not found we will push the element to the end.
+        if (index === -1) {
+          nodes.current = [...nodes.current, newItem];
+        } else {
+          nodes.current = [...nodes.current.slice(0, index), newItem, ...nodes.current.slice(index)];
```

Comment:
```suggestion
          nodes.current.splice(index, 0, newItem)
```

I'm pretty sure `splice` already handles what you want to do, unless you need to create a completely new array

Reply:
I remember there being a reason to reassign `nodes.current`, but I can't remember what it was, or if it still applies 😅. I'll switch it to `splice` for now, and switch it back if I remember or come across the issue again.